File: backend/app/schemas/person_profile.py
"""
Pydantic schemas for Person Profile models.
"""
from typing import Optional, List
from pydantic import BaseModel, EmailStr
from datetime import datetime, date
from enum import Enum


class PersonProfileVisibility(str, Enum):
    PRIVATE = "private"
    TENANT_SHARED = "tenant_shared"


# Base schemas
class PersonProfileBase(BaseModel):
    first_name: str
    last_name: Optional[str] = None
    display_name: Optional[str] = None
    notes: Optional[str] = None
    gender: Optional[str] = None
    is_active: Optional[bool] = True
    visibility: Optional[PersonProfileVisibility] = PersonProfileVisibility.PRIVATE


class PersonContactInfoBase(BaseModel):
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    address_line1: Optional[str] = None
    address_line2: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    postal_code: Optional[str] = None
    country: Optional[str] = None
    contact_type: Optional[str] = "personal"
    is_primary: Optional[bool] = False
    is_emergency_contact: Optional[bool] = False
    notes: Optional[str] = None
    visibility: Optional[PersonProfileVisibility] = PersonProfileVisibility.PRIVATE


class PersonProfessionalInfoBase(BaseModel):
    job_title: Optional[str] = None
    organization_id: Optional[int] = None
    organization_name: Optional[str] = None
    department: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    salary_range: Optional[str] = None
    work_location: Optional[str] = None
    employment_type: Optional[str] = None
    notes: Optional[str] = None
    is_current: Optional[bool] = True
    visibility: Optional[PersonProfileVisibility] = PersonProfileVisibility.PRIVATE


class PersonPersonalInfoBase(BaseModel):
    date_of_birth: Optional[date] = None
    anniversary_date: Optional[date] = None
    maiden_name: Optional[str] = None
    family_nickname: Optional[str] = None
    preferred_name: Optional[str] = None
    favorite_color: Optional[str] = None
    favorite_food: Optional[str] = None
    dietary_restrictions: Optional[str] = None
    allergies: Optional[str] = None
    personality_notes: Optional[str] = None
    interests_hobbies: Optional[str] = None
    notes: Optional[str] = None
    visibility: Optional[PersonProfileVisibility] = PersonProfileVisibility.PRIVATE


class PersonLifeEventBase(BaseModel):
    event_type: str
    title: str
    description: Optional[str] = None
    event_date: date
    location: Optional[str] = None
    my_role: Optional[str] = None
    significance: Optional[int] = None
    is_recurring: Optional[bool] = False
    notes: Optional[str] = None
    visibility: Optional[PersonProfileVisibility] = PersonProfileVisibility.PRIVATE


class PersonBelongingBase(BaseModel):
    name: str
    category: Optional[str] = None
    description: Optional[str] = None
    brand: Optional[str] = None
    model: Optional[str] = None
    estimated_value: Optional[str] = None
    acquisition_date: Optional[date] = None
    acquisition_method: Optional[str] = None
    relationship_context: Optional[str] = None
    notes: Optional[str] = None
    visibility: Optional[PersonProfileVisibility] = PersonProfileVisibility.PRIVATE
    is_active: Optional[bool] = True


class PersonRelationshipBase(BaseModel):
    person_a_id: int
    person_b_id: int
    relationship_type: str
    relationship_category: str
    relationship_strength: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    how_we_met: Optional[str] = None
    context: Optional[str] = None
    notes: Optional[str] = None
    is_mutual: Optional[bool] = True
    is_active: Optional[bool] = True
    visibility: Optional[PersonProfileVisibility] = PersonProfileVisibility.PRIVATE


# Create schemas
class PersonProfileCreate(PersonProfileBase):
    pass


class PersonContactInfoCreate(PersonContactInfoBase):
    person_id: int


class PersonProfessionalInfoCreate(PersonProfessionalInfoBase):
    person_id: int


class PersonPersonalInfoCreate(PersonPersonalInfoBase):
    person_id: int


class PersonLifeEventCreate(PersonLifeEventBase):
    person_id: int


class PersonBelongingCreate(PersonBelongingBase):
    person_id: int


class PersonRelationshipCreate(PersonRelationshipBase):
    pass


# Update schemas  
class PersonProfileUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    display_name: Optional[str] = None
    notes: Optional[str] = None
    gender: Optional[str] = None
    is_active: Optional[bool] = None
    visibility: Optional[PersonProfileVisibility] = None


class PersonContactInfoUpdate(BaseModel):
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    address_line1: Optional[str] = None
    address_line2: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    postal_code: Optional[str] = None
    country: Optional[str] = None
    contact_type: Optional[str] = None
    is_primary: Optional[bool] = None
    is_emergency_contact: Optional[bool] = None
    notes: Optional[str] = None
    visibility: Optional[PersonProfileVisibility] = None


class PersonProfessionalInfoUpdate(BaseModel):
    job_title: Optional[str] = None
    organization_id: Optional[int] = None
    organization_name: Optional[str] = None
    department: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    salary_range: Optional[str] = None
    work_location: Optional[str] = None
    employment_type: Optional[str] = None
    notes: Optional[str] = None
    is_current: Optional[bool] = None
    visibility: Optional[PersonProfileVisibility] = None


class PersonPersonalInfoUpdate(BaseModel):
    date_of_birth: Optional[date] = None
    anniversary_date: Optional[date] = None
    maiden_name: Optional[str] = None
    family_nickname: Optional[str] = None
    preferred_name: Optional[str] = None
    favorite_color: Optional[str] = None
    favorite_food: Optional[str] = None
    dietary_restrictions: Optional[str] = None
    allergies: Optional[str] = None
    personality_notes: Optional[str] = None
    interests_hobbies: Optional[str] = None
    notes: Optional[str] = None
    visibility: Optional[PersonProfileVisibility] = None


class PersonLifeEventUpdate(BaseModel):
    event_type: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    event_date: Optional[date] = None
    location: Optional[str] = None
    my_role: Optional[str] = None
    significance: Optional[int] = None
    is_recurring: Optional[bool] = None
    notes: Optional[str] = None
    visibility: Optional[PersonProfileVisibility] = None


class PersonBelongingUpdate(BaseModel):
    name: Optional[str] = None
    category: Optional[str] = None
    description: Optional[str] = None
    brand: Optional[str] = None
    model: Optional[str] = None
    estimated_value: Optional[str] = None
    acquisition_date: Optional[date] = None
    acquisition_method: Optional[str] = None
    relationship_context: Optional[str] = None
    notes: Optional[str] = None
    visibility: Optional[PersonProfileVisibility] = None
    is_active: Optional[bool] = None


class PersonRelationshipUpdate(BaseModel):
    relationship_type: Optional[str] = None
    relationship_category: Optional[str] = None
    relationship_strength: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    how_we_met: Optional[str] = None
    context: Optional[str] = None
    notes: Optional[str] = None
    is_mutual: Optional[bool] = None
    is_active: Optional[bool] = None
    visibility: Optional[PersonProfileVisibility] = None


# Response schemas (with database fields)
class PersonContactInfoInDBBase(PersonContactInfoBase):
    id: int
    person_id: int
    tenant_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PersonProfessionalInfoInDBBase(PersonProfessionalInfoBase):
    id: int
    person_id: int
    tenant_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PersonPersonalInfoInDBBase(PersonPersonalInfoBase):
    id: int
    person_id: int
    tenant_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PersonLifeEventInDBBase(PersonLifeEventBase):
    id: int
    person_id: int
    tenant_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PersonBelongingInDBBase(PersonBelongingBase):
    id: int
    person_id: int
    tenant_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PersonRelationshipInDBBase(PersonRelationshipBase):
    id: int
    tenant_id: int
    created_by_user_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Public API response schemas
class PersonContactInfo(PersonContactInfoInDBBase):
    pass


class PersonProfessionalInfo(PersonProfessionalInfoInDBBase):
    pass


class PersonPersonalInfo(PersonPersonalInfoInDBBase):
    pass


class PersonLifeEvent(PersonLifeEventInDBBase):
    pass


class PersonBelonging(PersonBelongingInDBBase):
    pass


class PersonRelationship(PersonRelationshipInDBBase):
    pass


class PersonProfileInDBBase(PersonProfileBase):
    id: int
    tenant_id: int
    created_by_user_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PersonProfile(PersonProfileInDBBase):
    # Include related information optionally
    contact_info: Optional[List[PersonContactInfo]] = None
    professional_info: Optional[List[PersonProfessionalInfo]] = None
    personal_info: Optional[List[PersonPersonalInfo]] = None
    life_events: Optional[List[PersonLifeEvent]] = None
    belongings: Optional[List[PersonBelonging]] = None


class PersonProfileInDB(PersonProfileInDBBase):
    pass


# Simplified response for lists
class PersonProfileSummary(BaseModel):
    id: int
    first_name: str
    last_name: Optional[str] = None
    display_name: Optional[str] = None
    visibility: PersonProfileVisibility
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/PersonProfile.ts
/**
 * TypeScript types for Person Profile (Relationship Management system)
 */

export enum PersonProfileVisibility {
  PRIVATE = "private",
  TENANT_SHARED = "tenant_shared"
}

// Base interfaces
export interface PersonProfileBase {
  first_name: string;
  last_name?: string;
  display_name?: string;
  notes?: string;
  gender?: string;
  is_active?: boolean;
  visibility?: PersonProfileVisibility;
}

export interface PersonContactInfoBase {
  email?: string;
  phone?: string;
  address_line1?: string;
  address_line2?: string;
  city?: string;
  state?: string;
  postal_code?: string;
  country?: string;
  contact_type?: string;
  is_primary?: boolean;
  is_emergency_contact?: boolean;
  notes?: string;
  visibility?: PersonProfileVisibility;
}

export interface PersonProfessionalInfoBase {
  job_title?: string;
  organization_id?: number;
  organization_name?: string;
  department?: string;
  start_date?: string; // ISO date string
  end_date?: string; // ISO date string
  salary_range?: string;
  work_location?: string;
  employment_type?: string;
  notes?: string;
  is_current?: boolean;
  visibility?: PersonProfileVisibility;
}

export interface PersonPersonalInfoBase {
  date_of_birth?: string; // ISO date string
  anniversary_date?: string; // ISO date string
  maiden_name?: string;
  family_nickname?: string;
  preferred_name?: string;
  favorite_color?: string;
  favorite_food?: string;
  dietary_restrictions?: string;
  allergies?: string;
  personality_notes?: string;
  interests_hobbies?: string;
  notes?: string;
  visibility?: PersonProfileVisibility;
}

export interface PersonLifeEventBase {
  event_type: string;
  title: string;
  description?: string;
  event_date: string; // ISO date string
  location?: string;
  my_role?: string;
  significance?: number;
  is_recurring?: boolean;
  notes?: string;
  visibility?: PersonProfileVisibility;
}

export interface PersonBelongingBase {
  name: string;
  category?: string;
  description?: string;
  brand?: string;
  model?: string;
  estimated_value?: string;
  acquisition_date?: string; // ISO date string
  acquisition_method?: string;
  relationship_context?: string;
  notes?: string;
  visibility?: PersonProfileVisibility;
  is_active?: boolean;
}

export interface PersonRelationshipBase {
  person_a_id: number;
  person_b_id: number;
  relationship_type: string;
  relationship_category: string;
  relationship_strength?: number;
  start_date?: string; // ISO date string
  end_date?: string; // ISO date string
  how_we_met?: string;
  context?: string;
  notes?: string;
  is_mutual?: boolean;
  is_active?: boolean;
  visibility?: PersonProfileVisibility;
}

// Create interfaces
export interface PersonProfileCreate extends PersonProfileBase {}

export interface PersonContactInfoCreate extends PersonContactInfoBase {
  person_id: number;
}

export interface PersonProfessionalInfoCreate extends PersonProfessionalInfoBase {
  person_id: number;
}

export interface PersonPersonalInfoCreate extends PersonPersonalInfoBase {
  person_id: number;
}

export interface PersonLifeEventCreate extends PersonLifeEventBase {
  person_id: number;
}

export interface PersonBelongingCreate extends PersonBelongingBase {
  person_id: number;
}

export interface PersonRelationshipCreate extends PersonRelationshipBase {}

// Update interfaces (all fields optional)
export interface PersonProfileUpdate {
  first_name?: string;
  last_name?: string;
  display_name?: string;
  notes?: string;
  gender?: string;
  is_active?: boolean;
  visibility?: PersonProfileVisibility;
}

export interface PersonContactInfoUpdate {
  email?: string;
  phone?: string;
  address_line1?: string;
  address_line2?: string;
  city?: string;
  state?: string;
  postal_code?: string;
  country?: string;
  contact_type?: string;
  is_primary?: boolean;
  is_emergency_contact?: boolean;
  notes?: string;
  visibility?: PersonProfileVisibility;
}

export interface PersonProfessionalInfoUpdate {
  job_title?: string;
  organization_id?: number;
  organization_name?: string;
  department?: string;
  start_date?: string;
  end_date?: string;
  salary_range?: string;
  work_location?: string;
  employment_type?: string;
  notes?: string;
  is_current?: boolean;
  visibility?: PersonProfileVisibility;
}

export interface PersonPersonalInfoUpdate {
  date_of_birth?: string;
  anniversary_date?: string;
  maiden_name?: string;
  family_nickname?: string;
  preferred_name?: string;
  favorite_color?: string;
  favorite_food?: string;
  dietary_restrictions?: string;
  allergies?: string;
  personality_notes?: string;
  interests_hobbies?: string;
  notes?: string;
  visibility?: PersonProfileVisibility;
}

export interface PersonLifeEventUpdate {
  event_type?: string;
  title?: string;
  description?: string;
  event_date?: string;
  location?: string;
  my_role?: string;
  significance?: number;
  is_recurring?: boolean;
  notes?: string;
  visibility?: PersonProfileVisibility;
}

export interface PersonBelongingUpdate {
  name?: string;
  category?: string;
  description?: string;
  brand?: string;
  model?: string;
  estimated_value?: string;
  acquisition_date?: string;
  acquisition_method?: string;
  relationship_context?: string;
  notes?: string;
  visibility?: PersonProfileVisibility;
  is_active?: boolean;
}

export interface PersonRelationshipUpdate {
  relationship_type?: string;
  relationship_category?: string;
  relationship_strength?: number;
  start_date?: string;
  end_date?: string;
  how_we_met?: string;
  context?: string;
  notes?: string;
  is_mutual?: boolean;
  is_active?: boolean;
  visibility?: PersonProfileVisibility;
}

// Response interfaces (with database fields)
export interface PersonContactInfo extends PersonContactInfoBase {
  id: number;
  person_id: number;
  tenant_id: number;
  created_at: string;
  updated_at?: string;
}

export interface PersonProfessionalInfo extends PersonProfessionalInfoBase {
  id: number;
  person_id: number;
  tenant_id: number;
  created_at: string;
  updated_at?: string;
}

export interface PersonPersonalInfo extends PersonPersonalInfoBase {
  id: number;
  person_id: number;
  tenant_id: number;
  created_at: string;
  updated_at?: string;
}

export interface PersonLifeEvent extends PersonLifeEventBase {
  id: number;
  person_id: number;
  tenant_id: number;
  created_at: string;
  updated_at?: string;
}

export interface PersonBelonging extends PersonBelongingBase {
  id: number;
  person_id: number;
  tenant_id: number;
  created_at: string;
  updated_at?: string;
}

export interface PersonRelationship extends PersonRelationshipBase {
  id: number;
  tenant_id: number;
  created_by_user_id: number;
  created_at: string;
  updated_at?: string;
}

export interface PersonProfile extends PersonProfileBase {
  id: number;
  tenant_id: number;
  created_by_user_id: number;
  created_at: string;
  updated_at?: string;
  
  // Optional related information
  contact_info?: PersonContactInfo[];
  professional_info?: PersonProfessionalInfo[];
  personal_info?: PersonPersonalInfo[];
  life_events?: PersonLifeEvent[];
  belongings?: PersonBelonging[];
}

// Simplified response for lists
export interface PersonProfileSummary {
  id: number;
  first_name: string;
  last_name?: string;
  display_name?: string;
  visibility: PersonProfileVisibility;
  created_at: string;
}

// API Response types
export interface ApiResponse<T> {
  data: T;
  message?: string;
}

export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  page: number;
  limit: number;
}

// Utility types for forms
export interface PersonProfileFormData extends PersonProfileCreate {
  // Additional form-specific fields can go here
}

export interface PersonContactInfoFormData extends PersonContactInfoCreate {
  // Additional form-specific fields can go here
}

// Error types
export interface ValidationError {
  field: string;
  message: string;
}

export interface ApiError {
  detail: string;
  validation_errors?: ValidationError[];
}